Issue mhmerrill#734: Create Registration List Function

pierce314159 opened the pull request:
This pull request (Issue #734) adds functionality to list all objects in the registry

- Adds `registered` field to the output of `ak.info()`
- Adds special `__RegisteredSymbols__` symbol which prints all registered objects when passed to `ak.info`  (`ak.info(ak.RegisteredSymbols)`).  This behaves similarly to the `__AllSymbols__` symbol
- Adds checks to return a message when info is called with an empty symbol table or `RegisteredSymbols` is used with an empty registry
- Updates `registration_test.py` to test new info() functionality

Change in behavior:  A `registered` field has been added to `ak.info()` . This may effect any logic which relies on the number or order of fields returned by `ak.info()`

glitch commented on src/MultiTypeSymbolTable.chpl:
```diff
@@ -320,19 +320,36 @@ module MultiTypeSymbolTable
+                        s += "name:%t dtype:%t size:%t ndim:%t shape:%t itemsize:%t registered:%t\n".format(n, 
+                                  dtype2str(tab.getBorrowed(n).dtype), tab.getBorrowed(n).size, 
+                                  tab.getBorrowed(n).ndim, tab.getBorrowed(n).shape, 
+                                  tab.getBorrowed(n).itemsize, registry.contains(n));
```
I think you did the right thing following suit from the previous code.  Two future optimizations we could probably make here are:
 1. Pull the item out of the map (`tab`) once vs the multiple `tab.getBorrowed(n)` calls.
 2. It looks like we're generating/formatting the same underlying string for AllSymbols, RegisteredSymbols, and single items.  If that's the case it might be worth it to consolidate that into its own function and just call it so it ends up looking something like:
 ```
for n in (registry|tab|etc.) {
  s += formatEntry(n, tab.getBorrowed(n));
}
...
proc formatEntry(name:string, item:GenSymEntry): string {
  return "name:%t dtype:%t size:%t ...".format(name, dtype2str(item.dtype), item.size ...);
}
```

If the strings are not in fact the same, forget I said anything :)
(No change requested from me.)

glitch commented on tests/registration_test.py:
```diff
@@ -125,3 +125,71 @@ def test_clear(self):
+        6. info(ak.RegisteredSymbols) only contains registered objects
+        '''
+        # Cleanup symbol table from previous tests
+        cleanup()
```
RE: cleanup() -> Nice catch here, the previous tests leave behind some state since the server & SymTab are shared across the tests.

mhmerrill commented on src/MultiTypeSymbolTable.chpl:
```diff
@@ -320,19 +320,36 @@ module MultiTypeSymbolTable
             var s: string;
-            if name == "__AllSymbols__" {
+            if tab.size == 0 {
+                s = "the symbol table is empty";
```
I think we may want to return a different string here maybe something like `"__EMPTY_SYMBOLTABLE__"`
what do you think? I also worry about the interaction between and empty symboltable/registry and your `cleanup()` function in the test flow.

hokiegeek2 replied:
@mhmerrill agreed, and perhaps an enum instead of a string?

mhmerrill commented on src/MultiTypeSymbolTable.chpl:
```diff
@@ -320,19 +320,36 @@ module MultiTypeSymbolTable
+            } 
+            else if name == "__RegisteredSymbols__" {
+                if registry.size == 0 {
+                    s = "the registry is empty";
```
same comment as above maybe return `"__EMPTY_REGISRTY__"`

mhmerrill commented on tests/registration_test.py:
```diff
@@ -125,3 +125,71 @@ def test_clear(self):
+
+def cleanup():
+    ak.clear()
+    for registered_object in ak.info(ak.AllSymbols).split('\n')[:-1]:
```
what happens when this gets called on an empty symboltable/registry?

pierce314159 replied:
I double checked this and `ak.info(ak.AllSymbols).split('\n')[:-1]` is an empty list when the symbol table is empty. So a `cleanup()` call with an empty registry should be the same as calling `ak.clear()`

mhmerrill replied:
ok, thanks

mhmerrill commented:
at some point we may want to convert the returned value from `ak.info()` to a JSON formatted thing so it can easily be a `dict` in python

mhmerrill reviewed (approved):
I think this looks good.

mhmerrill commented:
@pierce314159 I think this looks good.
Are there any commits outstanding from you? 
I also marked this ready for review.

pierce314159 commented:
In response to feedback above, I've

- [added a helper function](https://github.com/mhmerrill/arkouda/pull/743#discussion_r610754774) , `formatEntry(name:string, item:GenSymEntry)`, for the `ak.info` calls
- updated the `ak.info` doc string
- [changed the empty symbol table message](https://github.com/mhmerrill/arkouda/pull/743#discussion_r611727165)
- [changed the empty registry message](https://github.com/mhmerrill/arkouda/pull/743#discussion_r611727965)
- updated the `cleanup()` function in `registration_test.py` to be more clear and resilient to future changes

Future work for the `info` function will probably include moving towards json like @mhmerrill mentioned

pierce314159 commented:
> 
> 
> @pierce314159 I think this looks good.
> Are there any commits outstanding from you?
> I also marked this ready for review.

@mhmerrill, no outstanding commits! I was just doing a short write up of changes. This should be ready for review,

mhmerrill commented:
thanks! @pierce314159 i will merge when the CI finished cleanly ;-)
